### PR TITLE
Add Digit DP

### PR DESCRIPTION
**Pull request by sibasish14**
**Fixes issue:**

Fixes https://github.com/OpenGenus/cosmos/issues/3677

**Changes:**

This algorithm is used to count how many numbers x are there in the range a to b, where the digit d occurs exactly k times in x. This program uses a special type of dynamic programming approach called Digit DP. The implementation is done in C++.

### CODE REVIEW

**Comment by sibasish14** [comment]
@AdiChat please review.

File: code/dynamic_programming/src/digit_dp/digit_dp.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    else
+        return false;
+}
+
+int main()
+{
+    string a = "100", b = "100000000";
+    long ans = solve(b) - solve(a) + check(a);
+    cout << ans << "\n";
+
```
Comment (mansiag) [inline code/dynamic_programming/src/digit_dp/digit_dp.cpp]: Remove extra line

File: code/dynamic_programming/src/digit_dp/digit_dp.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    return dp[pos][count][f] = result;
+}
+
+long solve(string num)
+{
+    memset(dp, -1, sizeof(dp));
+    long result = digit_dp(num, 0, 0, 0);
+
```
Comment (mansiag) [inline code/dynamic_programming/src/digit_dp/digit_dp.cpp]: remove extra line

**Comment by sibasish14** [comment]
@mansiag required changes are done. Please check again.

File: code/dynamic_programming/src/digit_dp/digit_dp.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    // Try to place all the valid digits such that the number doesn't exceed b
+    for (long dgt = 0; dgt <= limit; dgt++)
+    {
+        long nf = f;
+        long ncount = count;
+
+        // The number is getting smaller at this position
+        if (f == 0 && dgt < limit)
+            nf=1;
```
Comment (mansiag) [inline code/dynamic_programming/src/digit_dp/digit_dp.cpp]: spaces

File: code/dynamic_programming/src/digit_dp/digit_dp.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        /* The number has already become smaller than b.
+         * We can place any digit now.
+         */
+        limit = 9;
+    }
+
+    // Try to place all the valid digits such that the number doesn't exceed b
+    for (long dgt = 0; dgt <= limit; dgt++)
+    {
+        long nf = f;
```
Comment (mansiag) [inline code/dynamic_programming/src/digit_dp/digit_dp.cpp]: better name than nf

File: code/dynamic_programming/src/digit_dp/digit_dp.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        long ncount = count;
```
Comment (mansiag) [inline code/dynamic_programming/src/digit_dp/digit_dp.cpp]: better name than ncount

**Comment by sibasish14** [comment]
@mansiag please review

**Comment by AdiChat** [comment]
@mansiag Thanks for reviewing this wonderful contribution, you are great :tada:

**Comment by AdiChat** [comment]
@sibasish14 Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**